test/test_type_promotion.py

```diff
@@ -138,39 +138,41 @@ def test_mixed_type_backward(self):
         f_dtypes = [torch.float, torch.double]
         f_dtypes = f_dtypes if self.device == 'cpu' else f_dtypes + [torch.half]
         i_dtypes = [torch.int, torch.long]
-        for f in ['add', 'sub', 'rsub', 'mul', 'div']:
+        for func in [torch.add, torch.sub, torch.rsub, torch.mul, torch.div]:
             for dtype1 in f_dtypes:
                 for dtype2 in (f_dtypes + i_dtypes):
                     x = torch.ones(10, requires_grad=True, dtype=dtype1, device=self.device)
                     y = torch.ones(10, dtype=dtype2, device=self.device)
-
-                    func = getattr(torch, f)
                     func(x, y).sum().backward()
 
-    # verifies that a.add(b) is the same as a.to(b.dtype).add(b) in cases
-    # where that should hold.
+    # verifies that torch.<op>(a, b) is the same as 
+    # torch.<op>(a.to(common_dtype), b.to(common_dtype)) in cases where that should hold.
     def test_many_promotions(self):
-        from_to = {
-            torch.float16: torch.float32,
-            torch.half: torch.float16,
-            torch.int: torch.long,
-            torch.uint8: torch.long,
-            torch.uint8: torch.float,
-            torch.int: torch.float,
-            torch.int: torch.double,
-            torch.int16: torch.long,
-            torch.float16: torch.double,
-            torch.bool: torch.long,
-            torch.bool: torch.float
-        }
-
-        for k, v in from_to.items():
-            a = torch.rand([3, 3], device=self.device).to(k)  # no _th_uniform for half on cpu.
-            b = torch.rand([3, 3], device=self.device).to(v)
-            c = a.add(b)
-            d = a.to(v).add(b)
-            self.assertEqual(c.dtype, d.dtype, message='from {} to {}'.format(k, v))
-            self.assertEqual(c, d, message='from {} to {}'.format(k, v))
+        for dt1 in torch.testing.get_all_math_dtypes(self.device):
+            for dt2 in torch.testing.get_all_math_dtypes(self.device):
+                for op in [torch.add, torch.sub, torch.mul, torch.div]:
+                    for contiguous in [True, False]:
+
+                        common_dtype = torch.promote_types(dt1, dt2)
+
+                        if op == torch.sub and common_dtype != torch.bool:
+                            # Subtraction, the `-` operator, with a bool tensor is not supported.
+                            continue
+
+                        a = torch.rand([20, 20], device=self.device).to(dt1)  # no _th_uniform for half on cpu.
+                        b = torch.rand([20, 20], device=self.device).to(dt2)
```

Comment:
We can make this do less work by including the dtype inside the ctor: `randn(..., dtype=dt2)`

Reply:
I had issues with that for half (see above comment # no _th_uniform for half on cpu.)

Reply:
Right, but we shouldn't be testing half on CPU, right?

Reply:
We do (even before this PR) test with half on CPU. It works because it gets promoted to a common type that has add kernels available. Eg, in the below, where we add a half tensor + float tensor, the addition is actually performed (inside tensor iterator) on promoted float tensors.

```
In [19]: torch.randn(5).to(torch.half).add(torch.ones(5, dtype=torch.float))
Out[19]: tensor([ 1.4626,  1.1921,  3.0762, -0.8486,  1.4751])
```

If alternatively we had attempted to add a scalar float to the half tensor it would fail, because we don't promote scalars within a kind, and so the common dtype would still be half inside TI.

Reply:
It wasn't intentional though that in switching from the list of explicit dtypes to get_all_math_dtypes I dropped the half testing. I think I should re-add half.

Reply:
I'm not sure what you mean, I don't know if we care to have _th_uniform work for half on CPU, and that could be an inefficient/undesirable op. My goal here is just to test that a half tensor gets correctly promoted, and don't want to use different constructors for different dtypes.

Reply:
I'm probably micro-optimizing here, but it is much faster to construct a tensor in one go with the correct dtype than to construct a tensor then use `.to()`. I am saying that if you think we should have `torch.rand` work for half support, we should file an issue and document it so that if we get torch.rand half support, we know to update these tests to run faster.

If you don't think we should have `torch.rand` work for half support, then this is fine.

Reply:
ok, I don't think half support is necessary, probably won't be used besides in unit tests.